### ISSUE TITLE
Learning Rate  (Adaptive Learning rate vs Learning Rate scheduler)  in Keras framework

### ISSUE THREAD

**Issue by vs74** [neuropoly/axondeepseg#217]
Learning rate implemented in tensorflow framework in AxonDeepSeg depending on the config file can be either one of the following:-
1. Polynomial learning rate decay(by default)
2. Exponential learning  rate decay
3. Constant learning rate

Since Keras is abstract level framework it does not provide with the option of training with polynomial, exponential decay. However, it provides with 2 options:

1. Adaptive Learning Rate ( which is having by default hyper tuned parameters )
2.  Learning Rate Scheduler ( to create your own custom learning rate function however just restricted to arguments such as epoch index and current learning rate as compared to the previous implementation having arguments like global_step, decay period, decay rate).

The adaptive Learning rate is generally better in terms of performance than Learning Rate Scheduler as it is generalized to large models and its parameters are hyper-tuned well as I have read in some articles (https://towardsdatascience.com/learning-rate-schedules-and-adaptive-learning-rate-methods-for-deep-learning-2c8f433990d1).

Now we have two options
1. To use the Adaptive Learning rate implementation and discard additional parameters in the config file for leaning rate.
2. Else use Learning Rate Scheduler but certainly modify the config file's parameters suited just to epoch index and learning rate.

How should we proceed with the above scenario?

**Comment by jcohenadad** [comment]
@perone can you please advise? Thanks!

**Comment by mathieuboudreau** [comment]
@vs74 if you're at the point that you can train some data with your Keras network, a good step would be to try both options and give us some details on how they differ in performance, accuracy, etc.

**Comment by vs74** [comment]
I was also thinking of comparing the results after implementing tensorboard for visualization of performance metrics. But one problem that Learning Rate Scheduler is that it's having own set of parameters which is different from the config file parameters for learning rate decay(tensorflow framework). How will I know the optimal value of parameters in Learning  Rate Scheduler to compare against Adaptive Learning rate?

**Comment by mathieuboudreau** [comment]
Not sure about optimal values, but any values that allow for a comparison is a good first step (and better than no comparison at all). Maybe that's a point that @perone could chime in, but I know that he's more familiar with PyTorch than he is with Keras

**Comment by mathieuboudreau** [comment]
Have you checked how/ifThibault implemented this learning rate in his Keras implementation of AxonDeepSeg? https://github.com/thibaulttabarin/UnAxSeg/

**Comment by mathieuboudreau** [comment]
Have you checked how/if @thibaulttabarin implemented the learning rate in his Keras implementation of AxonDeepSeg? https://github.com/thibaulttabarin/UnAxSeg

**Comment by perone** [comment]
Hi @vs74, the learning rate schedule isn't that important. However, if you want to use the polynomial decay, I've implemented it for Keras:

```
def poly_decay(epoch):
    initial_lrate = 0.001
    factor = 1.0 - (epoch / num_epoichs)
    lrate = initial_lrate * np.power(factor, 0.9)
    return lrate
```

For 800 epochs it looks like that:
![image](https://user-images.githubusercontent.com/412328/58912141-4f87bb80-8711-11e9-9951-ec8c1dafdd47.png)

You just need to adjust the `num_epochs` variable to your number of total epochs, the `initial_lrate` to the initial learning rate and the factor you want `0.9` for example. After that, just pass the function to the Keras `LearningRateScheduler`.

**Comment by vs74** [comment]
@mathieuboudreau, I have already looked into repo of @thibaulttabarin.  The repo contained the pre-trained model hence information regarding learning rate methods was hidden. Moreover, I didn't find any file for training the model so I couldn't figure out the implementation used for the learning rate.

**Comment by vs74** [comment]
@perone, Thanks for your reply. AxonDeepSeg is using polynomial decay and exponential decay for training the model. I was trying to replicate the learning rate decay methods in Keras framework. So I think we should use it. However the Keras framework uses an argument like `factor`  as you mentioned unlike the tensorflow( such as `global_step`, `decay period` and  decay rate`). So should we drop these parameters(`global_step`, `decay period` and  decay rate`) from config file and training script  for the conversion of AxonDeepSeg to Keras framework or else is there any way around?

**Comment by thibaulttabarin** [comment]
Hi,
that's right on my repo there is only the part to apply the model (user
version). I am currently finishing the training notebook. I should push it
by the end of the week.
Regarding the polydecay I found it wasn't necessary if you are using ADAM
optimizer... to my understanding the ADAM optimizer use an adaptative
learning rate which is what I have used.

cheers
T
